### PR TITLE
Build 2.12 (xhochy's branch)

### PR DESCRIPTION
**Pull request by xhochy**
Checklist
* [ ] Unvendor `re2`
* [x] Sort `TF_SYSTEM_LIBS` alphabetically
* [x] Loosen and update python package pins
* [x] unpin flatbuffers ~and absorb #320~ (no effect)
* [x] unvendor abseil
* [x] migrate libabseil 20230125
* [x] migrate libgrpc 1.54
* [x] migrate python 3.11

### CODE REVIEW

File: recipe/patches/0001-loosen-requirements.patch [hunk 1]
```diff
@@ -13,33 +13,26 @@ Co-authored-by: ngam
...
 --- a/tensorflow/tools/pip_package/setup.py
 +++ b/tensorflow/tools/pip_package/setup.py
-@@ -88,7 +88,6 @@ def standard_or_nightly(standard, nightly):
-     'gast >= 0.2.1, <= 0.4.0',
-     'google_pasta >= 0.1.1',
-     'h5py >= 2.9.0',
+@@ -92,7 +92,6 @@ def standard_or_nightly(standard, nightly):
+     # TODO(b/239052279): replace with external dependency on JAX repo once JAX
+     # no longer relies on TF.
+     'jax >= 0.3.15',
```
Comment (njzjz) [inline recipe/patches/0001-loosen-requirements.patch:22]: Do we need to add jax to the requirement?

File: recipe/patches/0001-loosen-requirements.patch [hunk 2]
```diff
@@ -13,33 +13,26 @@ Co-authored-by: ngam
...
 -    'libclang >= 13.0.0',
-     'numpy >= 1.20',
+     # TODO(b/263178356): numpy 1.24 breaks TF's tests
+     'numpy >= 1.22, <1.24',
```
Comment (njzjz) [inline recipe/patches/0001-loosen-requirements.patch:25]: We use numpy 1.21 which is inconsistent with this line.

File: recipe/patches/0001-loosen-requirements.patch [hunk 3]
```diff
@@ -13,33 +13,26 @@ Co-authored-by: ngam
...
-+    'protobuf >= 3.9.2',
-     'setuptools',
-     'six >= 1.12.0',
-     'termcolor >= 1.1.0',
-     'typing_extensions >= 3.6.6',
-     'wrapt >= 1.11.0',
+@@ -112,8 +111,6 @@ def standard_or_nightly(standard, nightly):
+     # TODO(b/266362323): wrapt==1.15.0rc0 incompatible with TF 2.12.0 RC0 (and
+     # nightly, but works with TF 2.11)
+     'wrapt >= 1.11.0, <1.15',
```
Comment (njzjz) [inline recipe/patches/0001-loosen-requirements.patch:30]: Do we need to add `<1.15`?

File: .ci_support/linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10.____cpython.yaml [hunk 4]
```diff
@@ -23,15 +23,15 @@ docker_image:
...
 - '72'
 libabseil:
-- '20220623.0'
+- '20230125'
 libcurl:
-- '7'
+- '8'
 libgrpc:
-- '1.51'
+- '1.52'
```
Comment (h-vetinari) [inline .ci_support/linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10.____cpython.yaml:34]: Would it be realistic to bump to grpc 1.54 here? That would be the "current" one before the (likely bigger change) protobuf 4

**Comment by xhochy** [comment]
@Tobias-Fischer Can you answer to @njzjz 's questions? These changes are all coming from your initial dependency update and I'm unsure why you have done them. I would probably revert them.

**Comment by Tobias-Fischer** [comment]
All of these changes simply reflect changes in the upstream requirements. If you look at the patch itself, it just removes two requirements (libclang and tensorflow-io-gcs-filesystem): https://github.com/xhochy/tensorflow-feedstock/blob/2.12-xhochy/recipe/patches/0001-loosen-requirements.patch The rest is just from the diff (numpy, jax, ...).

**Comment by xhochy** [comment]
> All of these changes simply reflect changes in the upstream requirements. If you look at the patch itself, it just removes two requirements (libclang and tensorflow-io-gcs-filesystem): https://github.com/xhochy/tensorflow-feedstock/blob/2.12-xhochy/recipe/patches/0001-loosen-requirements.patch The rest is just from the diff (numpy, jax, ...).

Ah, the Diff View is very confusing!

**Comment by xhochy** [comment]
We need to reflect these changes than also in the metadata. I'll take care of it later.

File: .ci_support/linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10.____cpython.yaml (hunk 4 above)
Comment (xhochy) [reply .ci_support/linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10.____cpython.yaml:34]: As mentioned in other threads, I would prefer to build both variants, 1.51 and 1.52.

Comment (h-vetinari) [reply .ci_support/linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10.____cpython.yaml:34]: > As mentioned in other threads, I would prefer to build both variants, 1.51 and 1.52.

Nothing supports 1.51 anymore. We [closed](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4370) the 1.52 migration 2 ago, and the 1.54 migration has effectively been done since that time as well (only missing tensorflow and a package that doesn't build anymore). In other words, going to 1.52 wouldn't help much with conflicts against other current packages.

In case that was perhaps confusing with the protobuf 4 situation, I'm not proposing to move to grpc >= 1.55 (which would require protobuf 4).

Comment (xhochy) [reply .ci_support/linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10.____cpython.yaml]: I was talking about 1.54 and 1.52 instead of 1.51 🙈 . I can then bump later to 1.54 and see whether it still passes.

Comment (h-vetinari) [reply .ci_support/linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10.____cpython.yaml]: OK cool, glad we double-checked. ;-)

To conclude my point though: Everything but tensorflow else has moved to 1.54. I would normally say "go ahead" if you want to build it twice also for 1.52 just to make sure, but considering the human (& compute) resources necessary to pull off a new version here, I would strongly suggest to skip 1.52.

Comment (Tobias-Fischer) [reply .ci_support/linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10.____cpython.yaml]: Somewhat related note - I think the libgrpc migrator should have migrated grpc_cpp as well. At the moment they are not in sync, but they should: https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/746eb797a786a605e6b5d0754dfcbbc27beba510/recipe/conda_build_config.yaml#L396-L398

Comment (h-vetinari) [reply .ci_support/linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10.____cpython.yaml]: No, it's unrelated. `grpc_cpp` was removed for 1.52, see [here](https://github.com/conda-forge/grpc-cpp-feedstock/issues/291). Where you're right is that the comment in the global pinning is now wrong, I'll go fix that.

**Comment by h-vetinari** [comment]
Can we also include the python 3.11 migrator from https://github.com/conda-forge/tensorflow-feedstock/pull/282?

File: recipe/patches/0001-loosen-requirements.patch (hunk 2 above)
Comment (h-vetinari) [reply recipe/patches/0001-loosen-requirements.patch:25]: Lower build-time dependency is usually fine. The runtime dependence is a bigger question here, but unless the breakage is severe (is there an upstream issue where this is documented?), I'd like to leave numpy unpinned.

File: recipe/patches/0001-loosen-requirements.patch (hunk 3 above)
Comment (h-vetinari) [reply recipe/patches/0001-loosen-requirements.patch:30]: We could avoid this if we backport https://github.com/tensorflow/tensorflow/pull/60641 (AFAICT from looking around upstream, where the pin was lifted again inhttps://github.com/tensorflow/tensorflow/pull/60661).

**Comment by njzjz** [comment]
`linux_64_c_compiler_version10cuda_compiler_version11.2cxx_compiler_version10python3.10` reports an error:

```
[16,115 / 18,695] Compiling llvm/lib/Transforms/Instrumentation/CGProfile.cpp; 3s local ... (2 actions running)
ERROR: /home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/work/tensorflow/compiler/xla/stream_executor/cuda/BUILD:519:11: Compiling tensorflow/compiler/xla/stream_executor/cuda/cuda_graph.cc failed: (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command 
  (cd /home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_build_env/share/bazel/e114260324380ce1d789ab922194b40a/execroot/org_tensorflow && \
  exec env - \
    CUDA_TOOLKIT_PATH=/usr/local/cuda-11.2 \
    GCC_HOST_COMPILER_PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_build_env/bin/x86_64-conda-linux-gnu-gcc \
    PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/work:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/condabin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_build_env/bin:/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin:/opt/conda/bin:/opt/conda/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/conda/bin:/usr/local/cuda/bin \
    PWD=/proc/self/cwd \
    PYTHON_BIN_PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/bin/python \
    PYTHON_LIB_PATH=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/python3.10/site-packages \
    TF2_BEHAVIOR=1 \
    TF_CUDA_COMPUTE_CAPABILITIES=sm_35,sm_50,sm_60,sm_62,sm_70,sm_72,sm_75,sm_80,sm_86,compute_86 \
    TF_CUDA_PATHS=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac,/usr/local/cuda \
    TF_CUDA_VERSION=11.2 \
    TF_CUDNN_VERSION=8 \
    TF_NCCL_VERSION=2.18 \
    TF_SYSTEM_LIBS=absl_py,astor_archive,astunparse_archive,boringssl,com_github_googlecloudplatform_google_cloud_cpp,com_github_grpc_grpc,com_google_protobuf,curl,cython,dill_archive,flatbuffers,gast_archive,gif,icu,libjpeg_turbo,org_sqlite,png,pybind11,snappy,zlib \
  custom_toolchain/crosstool_wrapper_driver_is_not_gcc -MD -MF bazel-out/k8-opt/bin/tensorflow/compiler/xla/stream_executor/cuda/_objs/cuda_graph/cuda_graph.pic.d '-frandom-seed=bazel-out/k8-opt/bin/tensorflow/compiler/xla/stream_executor/cuda/_objs/cuda_graph/cuda_graph.pic.o' -fPIC -DEIGEN_MPL2_ONLY '-DEIGEN_MAX_ALIGN_BYTES=64' -DTF_USE_SNAPPY -iquote . -iquote bazel-out/k8-opt/bin -iquote external/com_google_absl -iquote bazel-out/k8-opt/bin/external/com_google_absl -iquote external/local_config_cuda -iquote bazel-out/k8-opt/bin/external/local_config_cuda -iquote external/eigen_archive -iquote bazel-out/k8-opt/bin/external/eigen_archive -iquote external/nsync -iquote bazel-out/k8-opt/bin/external/nsync -iquote external/double_conversion -iquote bazel-out/k8-opt/bin/external/double_conversion -iquote external/com_google_protobuf -iquote bazel-out/k8-opt/bin/external/com_google_protobuf -iquote external/snappy -iquote bazel-out/k8-opt/bin/external/snappy -iquote external/com_googlesource_code_re2 -iquote bazel-out/k8-opt/bin/external/com_googlesource_code_re2 -iquote external/farmhash_archive -iquote bazel-out/k8-opt/bin/external/farmhash_archive -Ibazel-out/k8-opt/bin/external/local_config_cuda/cuda/_virtual_includes/cuda_headers_virtual -isystem external/local_config_cuda/cuda -isystem bazel-out/k8-opt/bin/external/local_config_cuda/cuda -isystem external/local_config_cuda/cuda/cuda/include -isystem bazel-out/k8-opt/bin/external/local_config_cuda/cuda/cuda/include -isystem third_party/eigen3/mkl_include -isystem bazel-out/k8-opt/bin/third_party/eigen3/mkl_include -isystem external/eigen_archive -isystem bazel-out/k8-opt/bin/external/eigen_archive -isystem external/nsync/public -isystem bazel-out/k8-opt/bin/external/nsync/public -isystem external/farmhash_archive/src -isystem bazel-out/k8-opt/bin/external/farmhash_archive/src -isystem /home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include '-march=nocona' '-mtune=haswell' -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include '-fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/work=/usr/local/src/conda/tensorflow-split-2.12.0' '-fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac=/usr/local/src/conda-prefix' -isystem /usr/local/cuda/include -DNDEBUG '-D_FORTIFY_SOURCE=2' -O2 -isystem /home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include -isystem /usr/local/cuda/include -fvisibility-inlines-hidden '-std=c++17' '-fmessage-length=0' '-march=nocona' '-mtune=haswell' -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include '-fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/work=/usr/local/src/conda/tensorflow-split-2.12.0' '-fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac=/usr/local/src/conda-prefix' -isystem /usr/local/cuda/include -DNDEBUG '-D_FORTIFY_SOURCE=2' -O2 -isystem /home/conda/feedstock_root/build_artifacts/tensorflow-split_1688120916119/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/include -isystem /usr/local/cuda/include -Wno-all -Wno-extra -Wno-deprecated -Wno-deprecated-declarations -Wno-ignored-attributes -Wno-array-bounds -Wunused-result '-Werror=unused-result' -Wswitch '-Werror=switch' '-Wno-error=unused-but-set-variable' -DAUTOLOAD_DYNAMIC_KERNELS '-std=c++17' -c tensorflow/compiler/xla/stream_executor/cuda/cuda_graph.cc -o bazel-out/k8-opt/bin/tensorflow/compiler/xla/stream_executor/cuda/_objs/cuda_graph/cuda_graph.pic.o)
# Configuration: 95f1cd756c5b200e444b11560d9e13b428636da3bb090a71139de5423db5c03a
# Execution platform: @local_execution_config_platform//:platform
tensorflow/compiler/xla/stream_executor/cuda/cuda_graph.cc: In function 'tsl::StatusOr<stream_executor::gpu::OwnedCudaGraph> stream_executor::gpu::CaptureCudaGraph(stream_executor::Stream*, absl::lts_20220623::AnyInvocable<tsl::Status()>, cudaStreamCaptureMode)':
tensorflow/compiler/xla/stream_executor/cuda/cuda_graph.cc:135:21: error: 'cudaGraphDebugDotFlagsVerbose' was not declared in this scope
  135 |         int flags = cudaGraphDebugDotFlagsVerbose;
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tensorflow/compiler/xla/stream_executor/cuda/cuda_graph.cc:136:24: error: 'cudaGraphDebugDotPrint' was not declared in this scope
  136 |         if (auto err = cudaGraphDebugDotPrint(graph, file.c_str(), flags);
      |                        ^~~~~~~~~~~~~~~~~~~~~~
INFO: Elapsed time: 19906.085s, Critical Path: 372.32s
```

`#if CUDA_VERSION >= 12000`, found in v2.13.0-rc0, is missing in v2.12.0. I‘ll add a patch to fix it.

**Comment by xhochy** [comment]
Building `tensorflow-estimator` is failing for me with:

```
Packaging tensorflow-estimator
INFO:conda_build.build:Packaging tensorflow-estimator
Reloading output folder: /home/uwe/mambaforge/conda-bld
home/uwe/mambaforge/conda-bld/linux-64               3.9kB @  57.2MB/s  0.0s
home/uwe/mambaforge/conda-bld/noarch               127.0 B @   5.3MB/s  0.0s
Mamba failed to solve:
 - libstdcxx-ng 13.1.0.* hfd8a6a1_0
 - _libgcc_mutex 0.1.* conda_forge
 - ca-certificates 2023.5.7.* hbcca054_0
 - ld_impl_linux-64 2.40.* h41732ed_0
 - tzdata 2023c h71feb2d_0
 - libgomp 13.1.0.* he5830b7_0
 - _openmp_mutex 4.5.* 2_gnu
 - libgcc-ng 13.1.0.* he5830b7_0
 - bzip2 1.0.8.* h7f98852_4
 - libffi 3.4.2.* h7f98852_5
 - libnsl 2.0.0.* h7f98852_0
 - libuuid 2.38.1.* h0b41bf4_0
 - libzlib 1.2.13.* hd590300_5
 - ncurses 6.4.* hcb278e6_0
 - openssl 3.1.1.* hd590300_1
 - xz 5.2.6.* h166bdaf_0
 - libsqlite 3.42.0.* h2797004_0
 - readline 8.2.* h8228510_1
 - tk 8.6.12.* h27826a3_0
 - python 3.10.12.* hd12c33a_0_cpython
 - packaging 23.1.* pyhd8ed1ab_0
 - setuptools 68.0.0.* pyhd8ed1ab_0
 - wheel 0.40.0.* pyhd8ed1ab_0
 - pip 23.1.2.* pyhd8ed1ab_0
 - tensorflow-base 2.12.0.* cpu_py310h854a703_0

with channels:

The reported errors are:
- Encountered problems while solving:
-   - package tensorflow-base-2.12.0-cpu_py310h854a703_0 requires libabseil >=20230125.3,<20230126.0a0, but none of the providers can be installed
```

**Comment by xhochy** [comment]
`micromamba` gives a better info:

```
error    libmamba Could not solve for environment specs
    The following packages are incompatible
    ├─ python 3.10.12.* hd12c33a_0_cpython is installable and it requires
    │  └─ python_abi 3.10.* *_cp310, which can be installed;
    └─ tensorflow-base 2.12.0.* cpu_py310h854a703_0 is not installable because it requires
       ├─ grpcio 1.52.*  with the potential options
       │  ├─ grpcio [1.52.0|1.52.1] would require
       │  │  └─ libgrpc [1.52.0 hbc085c6_0|1.52.1 hbc085c6_0|1.52.1 hcf146ea_1], which requires
       │  │     └─ libabseil 20230125 cxx17*, which can be installed;
       │  ├─ grpcio [1.52.0|1.52.1] would require
       │  │  └─ python_abi 3.11.* *_cp311, which conflicts with any installable versions previously reported;
       │  ├─ grpcio [1.52.0|1.52.1] would require
       │  │  └─ python_abi 3.8.* *_cp38, which conflicts with any installable versions previously reported;
       │  ├─ grpcio [1.52.0|1.52.1] would require
       │  │  └─ python_abi 3.9.* *_cp39, which conflicts with any installable versions previously reported;
       │  └─ grpcio 1.52.1 would require
       │     └─ libgrpc 1.52.1 hb20ce57_2, which requires
       │        └─ libabseil 20230125.2 cxx17*, which can be installed;
       └─ libabseil >=20230125.3,<20230126.0a0 , which conflicts with any installable versions previously reported.
critical libmamba Could not solve for environment specs
```

There is one remaining `grpc_cpp` usage in the recipe.

**Comment by xhochy** [comment]
We have some issues with the mixing of various Abseil versions. I'm working on a patch locally, please don't push here.

**Comment by hmaarrfk** [comment]
Some runtime pinnings need to be adjusted. I was able to build the CPU package but got stuck at:

```
+ pip check
tensorflow 2.12.0 requires jax, which is not installed.
tensorflow 2.12.0 has requirement numpy<1.24,>=1.22, but you have numpy 1.25.0.
tensorflow 2.12.0 has requirement wrapt<1.15,>=1.11.0, but you have wrapt 1.15.0.
Tests failed for tensorflow-estimator-2.12.0-cuda112py310ha5e6de5_0.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken

```

**Comment by xhochy** [comment]
> Some runtime pinnings need to be adjusted. I was able to build the CPU package but got stuck at:
> 
> 
> 
> ```
> 
> + pip check
> 
> tensorflow 2.12.0 requires jax, which is not installed.
> 
> tensorflow 2.12.0 has requirement numpy<1.24,>=1.22, but you have numpy 1.25.0.
> 
> tensorflow 2.12.0 has requirement wrapt<1.15,>=1.11.0, but you have wrapt 1.15.0.
> 
> Tests failed for tensorflow-estimator-2.12.0-cuda112py310ha5e6de5_0.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
> 
> 
> 
> ```

Already on my list!

**Comment by h-vetinari** [comment]
> Some runtime pinnings need to be adjusted.

These were what @njzjz was mentioning above. The compatibility patch for wrapt is already backported in this PR. I don't know what issues there are with numpy, but I [proposed](https://github.com/conda-forge/tensorflow-feedstock/pull/323#discussion_r1247742975) not to cap.

**Review by h-vetinari — COMMENTED** [review]
We should probably also do:
```diff
-  - flatbuffers >=2
+  - flatbuffers
```
& rerender, to gain some metadata transparency. We're already pulling in the latest builds (so effectively, using https://github.com/conda-forge/tensorflow-feedstock/pull/320); luckily flatbuffers [has](https://github.com/conda-forge/flatbuffers-feedstock/blob/main/recipe/meta.yaml) a run-export so things shouldn't be too broken.

Also note that flatbuffers switched from v2.0.8 to a CalVer-style v22.9.24 that's been the rule of the game since.

**Comment by h-vetinari** [comment]
> Is there an issue discussing the numpy cap?

I asked the same thing above. The comment in the upstream file references an internal ID, and just says tests fail with 1.24. This is IMO not good enough to justify the cap - yes, newer numpy may break some corner cases in tests, but I strongly doubt that the package is unusable. I think it would be _less_ usable with a <1.24 numpy cap; 1.23 is over a year old.

File: recipe/patches/0007-Update-Abseil-to-the-latest-LTS-replace-or-with-in-d.patch [hunk 5]
```diff
@@ -0,0 +1,450 @@
+From 96cd7d051fa6e0df83efcee259efec1092faf1c3 Mon Sep 17 00:00:00 2001
+From: David Dunleavy <ddunleavy@google.com>
+Date: Thu, 13 Apr 2023 10:02:13 -0700
+Subject: [PATCH 7/9] Update Abseil to the latest LTS, replace `or` with `||`
+ in data_format_ops.cc
+
+Current build requires a more recent Abseil with AnyInvocable, otherwise the build raise the following error:
```
Comment (h-vetinari) [inline recipe/patches/0007-Update-Abseil-to-the-latest-LTS-replace-or-with-in-d.patch:7]: Could you leave a comment in `meta.yaml` where this commit comes from?

File: recipe/build.sh [hunk 6]
```diff
@@ -38,6 +38,7 @@ export TF_SYSTEM_LIBS="
   astor_archive
   astunparse_archive
   boringssl
+  com_google_absl
   com_github_googlecloudplatform_google_cloud_cpp
   com_github_grpc_grpc
   com_google_protobuf
```
Comment (h-vetinari) [inline recipe/build.sh:44]: This list was alphabetical so far, please sort `com_google_absl` at the appropriate place.

Also, seeing the comment just above `TF_SYSTEM_LIBS`, we now have C++17 re2 in conda-forge.

Comment (xhochy) [reply recipe/build.sh:44]: Added both things as tasks in the top comment. I will check that once a clean build with `com_google_absl` has passed.

File: recipe/meta.yaml [hunk 7]
```diff
@@ -367,12 +380,13 @@ outputs:
         - {{ compiler('cuda') }}  # [cuda_compiler_version != "None"]
       # host requirements to pick up run_exports
       host:
-        - libabseil
+        - flatbuffers
```
Comment (h-vetinari) [inline recipe/meta.yaml:383]: Global hostdeps still use 
```
    - flatbuffers >=2
```

File: recipe/patches/0007-Update-Abseil-to-the-latest-LTS-replace-or-with-in-d.patch (hunk 5 above)
Comment (xhochy) [reply recipe/patches/0007-Update-Abseil-to-the-latest-LTS-replace-or-with-in-d.patch:7]: I couldn't find a PR that matches it; thus I didn't add a link.

File: recipe/meta.yaml (hunk 7 above)
Comment (xhochy) [reply recipe/meta.yaml:383]: Unpinning it now.

**Comment by xhochy** [comment]
If someone could take care of changing the Python pins (and push here), this would be awesome. I'm taking care of the rest.

**Comment by h-vetinari** [comment]
> If someone could take care of changing the Python pins (and push here), this would be awesome. I'm taking care of the rest.

Are you agreed with:
* add jax as dep
* keep numpy unpinned (and patch out the pin)
* keep wrapt unpinned (after we backported the compat patch)
* remove `>=2` from flatbuffers
?

Happy to do that.

**Comment by h-vetinari** [comment]
> This should be already part of the build?

That's what I meant.

File: recipe/patches/0007-Update-Abseil-to-the-latest-LTS-replace-or-with-in-d.patch (hunk 5 above)
Comment (h-vetinari) [reply recipe/patches/0007-Update-Abseil-to-the-latest-LTS-replace-or-with-in-d.patch:7]: Added a comment about the commit.

**Comment by h-vetinari** [comment]
So I tried to carry #320 and rerender, but smithy will not pick up a flatbuffers pin for some reasons 🤷

File: recipe/patches/0001-loosen-requirements.patch (hunk 2 above)
Comment (h-vetinari) [reply recipe/patches/0001-loosen-requirements.patch]: Done.

File: recipe/patches/0001-loosen-requirements.patch (hunk 1 above)
Comment (h-vetinari) [reply recipe/patches/0001-loosen-requirements.patch:22]: Done.

**Comment by xhochy** [comment]
> So I tried to carry #320 and rerender, but smithy will not pick up a flatbuffers pin for some reasons 🤷

See https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4637

**Comment by xhochy** [comment]
Enabling `com_googlesource_code_re2` currently fails with the following. I would omit it for now.

```
# Configuration: 724b03632281a6986c71d0dcc81f49dd3e81c6d4d13af369084afb77570e5bda
# Execution platform: @local_execution_config_platform//:platform
tensorflow/core/data/service/url.cc: In member function 'void tensorflow::data::URL::Parse(absl::lts_20230125::string_view)':
tensorflow/core/data/service/url.cc:32:22: error: no matching function for call to 're2::RE2::FullMatch(absl::lts_20230125::string_view&, absl::lts_20230125::string_view&, std::string*, std::string*)'
   32 |   if (!RE2::FullMatch(url, regexp, &host_, &port_)) {
      |        ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./tensorflow/tsl/platform/regexp.h:26,
                 from ./tensorflow/core/platform/regexp.h:18,
                 from tensorflow/core/data/service/url.cc:20:
/home/uwe/mambaforge/conda-bld/tensorflow-split_1688452071854/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include/re2/re2.h:398:15: note: candidate: 'template<class ... A> static bool re2::RE2::FullMatch(const re2::StringPiece&, const re2::RE2&, A&& ...)'
  398 |   static bool FullMatch(const StringPiece& text, const RE2& re, A&&... a) {
      |               ^~~~~~~~~
/home/uwe/mambaforge/conda-bld/tensorflow-split_1688452071854/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include/re2/re2.h:398:15: note:   template argument deduction/substitution failed:
tensorflow/core/data/service/url.cc:32:28: note:   cannot convert 'regexp' (type 'absl::lts_20230125::string_view' {aka 'std::basic_string_view<char>'}) to type 'const re2::RE2&'
   32 |   if (!RE2::FullMatch(url, regexp, &host_, &port_)) {
      |                            ^~~~~~
INFO: Elapsed time: 343.539s, Critical Path: 120.78s
INFO: 600 processes: 345 internal, 255 local.
```

**Comment by xhochy** [comment]
linux-cpu and osx-arm64 passed locally, waiting for linux-cuda to build but I think this is ready for a review. I can kick off the full matrix build later today/tomorrow morning.

File: recipe/meta.yaml [hunk 8]
```diff
@@ -256,7 +266,8 @@ outputs:
         - cat test_tensorflow.py     # [unix]
         - python test_tensorflow.py  # [build_platform == target_platform]
         - tf_upgrade_v2 --help       # [build_platform == target_platform]
-        - saved_model_cli --help     # [build_platform == target_platform]
+        # --help exits with exit code 1
+        - test -f $PREFIX/bin/saved_model_cli  # [build_platform == target_platform]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:270]:
```suggestion
        - test -f $PREFIX/bin/saved_model_cli  # [unix]
```

**Comment by h-vetinari** [comment]
> Enabling `com_googlesource_code_re2` currently fails with the following. I would omit it for now.

The fact that the symbol shows `absl::string_view` and not `std::string_view` probably points to C++17 not being applied correctly somewhere. 

But the re2 thing is not urgent

**Comment by njzjz** [comment]
[v2.12.1](https://github.com/tensorflow/tensorflow/releases/tag/v2.12.1) was just released. Do we need to bump the version?

**Comment by xhochy** [comment]
Just in time when the Linux builds finished on my machines 🤣 I can update the patches and rebuild with 2.12.1 if it is a simple rebase.

I'm struggling to find the time for OSX builds at the moment. Can someone else volunteer for them?

**Comment by xhochy** [comment]
Tackling 2.12.1 over here: https://github.com/conda-forge/tensorflow-feedstock/pull/325
